scrapy/core/spidermw.py

```diff
@@ -24,12 +49,10 @@ def _add_middleware(self, mw):
         super(SpiderMiddlewareManager, self)._add_middleware(mw)
         if hasattr(mw, 'process_spider_input'):
             self.methods['process_spider_input'].append(mw.process_spider_input)
-        if hasattr(mw, 'process_spider_output'):
-            self.methods['process_spider_output'].insert(0, mw.process_spider_output)
-        if hasattr(mw, 'process_spider_exception'):
-            self.methods['process_spider_exception'].insert(0, mw.process_spider_exception)
         if hasattr(mw, 'process_start_requests'):
             self.methods['process_start_requests'].insert(0, mw.process_start_requests)
+        self.methods['process_spider_output'].insert(0, getattr(mw, 'process_spider_output', None))
+        self.methods['process_spider_exception'].insert(0, getattr(mw, 'process_spider_exception', None))
```

Comment:
Are you adding None to the list to make start_index (and slicing) work properly?